drivers/led_strip/ws2812_rpi_pico_pio.c

```diff
@@ -0,0 +1,194 @@
+	uint32_t frequency;
+	const uint8_t *color_mapping;
+	uint16_t reset_delay;
+};
+
+/*
+ * This pio program runs 10 cycles per 1 loop.
+ * The `out` instruction output 000 to the sideset pin.
+ * The second `jmp` instruction output 11 to sideset pin.
+ * This `jmp` instruction jump to line 3 if the value of register x is true,
+ * otherwise jump to line 4.
+ * The third `jmp` instruction outputs 11111 to the sideset pin.
+ * After outputting, return to firstline.
+ * The fourth `nop` instruction outputs 00000. After outputting, return to firstline.
+ *
+ * The final output is 0001100000 in case of register x is false.
+ * It represents code 0, defined in the datasheet.
+ * And outputs 0001111111 in case of x is true. It represents code 1.
+ */
+RPI_PICO_PIO_DEFINE_PROGRAM(ws2812, 0, 3,
+		/*  .wrap_target                          */
+	0x6221, /*  0: out    x, 1            side 0 [2]  */
+	0x1123, /*  1: jmp    !x, 3           side 1 [1]  */
+	0x1400, /*  2: jmp    0               side 1 [4]  */
+	0xa442, /*  3: nop                    side 0 [4]  */
+		/*  .wrap                                 */
+);
+
+static void ws2812_program_init(PIO pio, uint32_t sm, uint32_t pin, float freq)
+{
+	float clkdiv = sys_clock_hw_cycles_per_sec() / (CYCLES_PER_BIT * freq);
+	uint32_t offset;
+
+	offset = pio_add_program(pio, &ws2812_program);
+
+	pio_sm_config sm_config = pio_get_default_sm_config();
+
+	sm_config_set_wrap(&sm_config, offset + RPI_PICO_PIO_GET_WRAP_TARGET(ws2812),
+			   offset + RPI_PICO_PIO_GET_WRAP(ws2812));
+	sm_config_set_sideset(&sm_config, 1, false, false);
+	sm_config_set_sideset_pins(&sm_config, pin);
+	sm_config_set_out_shift(&sm_config, false, true, 24);
+	sm_config_set_fifo_join(&sm_config, PIO_FIFO_JOIN_TX);
+	sm_config_set_clkdiv(&sm_config, clkdiv);
+	pio_sm_set_consecutive_pindirs(pio, sm, pin, 1, true);
+	pio_sm_init(pio, sm, offset, &sm_config);
+	pio_sm_set_enabled(pio, sm, true);
+}
+
+/*
+ * Latch current color values on strip and reset its state machines.
+ */
+static inline void ws2812_reset_delay(uint16_t delay)
+{
+	k_usleep(delay);
+}
+
+static int ws2812_rpi_pico_pio_update_rgb(const struct device *dev, struct led_rgb *pixels,
+					  size_t num_pixels)
+{
+	const struct ws2812_rpi_pico_pio_config *config = dev->config;
+	struct ws2812_rpi_pico_pio_data *data = dev->data;
+	PIO pio = pio_rpi_pico_get_pio(config->piodev);
+
+	for (size_t i = 0; i < num_pixels; i++) {
+		uint32_t color = 0;
+
+		for (size_t j = 0; j < config->num_colors; j++) {
+			switch (config->color_mapping[j]) {
+			/* White channel is not supported by LED strip API. */
+			case LED_COLOR_ID_WHITE:
+				color |= 0;
+				break;
+			case LED_COLOR_ID_RED:
+				color |= pixels[i].r << (8 * (2 - j));
+				break;
+			case LED_COLOR_ID_GREEN:
+				color |= pixels[i].g << (8 * (2 - j));
+				break;
+			case LED_COLOR_ID_BLUE:
+				color |= pixels[i].b << (8 * (2 - j));
+				break;
+			}
+		}
+
+		pio_sm_put_blocking(pio, data->sm, color << 8);
```

Comment:
How is handled this third byte set to 0 ? Won't it be interpreted by the next controller in the strip as the value of its first color ?

Reply:
In fact, it won't. Let me clarify my thinking: do we emit extra zeros on the wire (one byte out of four) ? And if so, can we optimize the output data in order to avoid that ?

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/55226/files#diff-e3c7e6a4007d7de0939e6d3409d542a944cd2771cb6db37fb806945f3989a257R56

The process at this line enforcing uses only 24bit for actual transmission.
The last 1 byte will discard.

Reply:
Improvement:
Detect 24bit/32bit color from the `color-mapping` configuration, applying it to sending bits.